server/src/test/java/org/elasticsearch/index/snapshots/blobstore/BlobStoreIndexShardSnapshotsTest.java

```diff
@@ -0,0 +1,175 @@
+        in.setVersion(org.elasticsearch.Version.V_5_10_1);
+        fromStream = BlobStoreIndexShardSnapshots.fromStream(in);
+        assertThat(isSame(blobStoreIndexShardSnapshots, fromStream)).isTrue();
+
+    }
+
+    @Test
+    public void test_read_from_stream_allocations() throws Exception {
+        BytesReference bytesReference = INDEX_SHARD_SNAPSHOTS_FORMAT.serialize(prepareData(), "dummyBlobName", true);
+
+        ThreadMXBean threadMXBean = ManagementFactory.getPlatformMXBean(ThreadMXBean.class);
+        long threadId = Thread.currentThread().threadId();
+        long allocatedBytesBegin = threadMXBean.getThreadAllocatedBytes(threadId);
+
+        INDEX_SHARD_SNAPSHOTS_FORMAT.deserialize(
+            "dummyBlobname",
+            writableRegistry(),
+            xContentRegistry(),
+            bytesReference
+        );
+
+        long allocatedBytesAfter = threadMXBean.getThreadAllocatedBytes(threadId);
+        long allocatedBytes = allocatedBytesAfter - allocatedBytesBegin;
+        // Actual value fluctuates around 9_877_672.
+        // Used to fluctuate around 29_228_160 on 5.10.2 when instances were not re-used.
+        assertThat(allocatedBytes).isLessThanOrEqualTo(10_000_000L);
+    }
+
+    /**
+     * Returns an instance with SnapshotFiles instances having their indexFiles overlapping
+     * and those overlaps being re-used as a same instance.
+     *
+     * We need to ensure that we can read such optimal file in a way that deserialized instance also has an optimal structure.
+     */
+    private static BlobStoreIndexShardSnapshots prepareData() throws Exception {
+        // Method getThreadAllocatedBytes is a rough estimation and returned value fluctuates a little bit.
+        // Generate a bigger list so that difference with/without the fix is visible (like 3x) and can't be treated as a noise.
+        List<FileInfo> fileInfos = new ArrayList<>();
+        for (int i = 1; i <= 10_000; i++) {
+            String name = "name" + i;
+            fileInfos.add(
+                new FileInfo(
+                    name,
+                    new StoreFileMetadata(name, 1, "dummy_checksum", org.apache.lucene.util.Version.LUCENE_9_12_0),
+                    new ByteSizeValue(i)
+                )
+            );
+        }
+
+        // Different SnapshotFiles can refer to the same FileInfo-s in the SnapshotFiles.indexFiles field.
+        // Making 2 lists almost completely overlapping to make difference visible and eliminate getThreadAllocatedBytes fluctuation noise.
+        SnapshotFiles snapshot1 = new SnapshotFiles("1", fileInfos.subList(10, fileInfos.size()), null);
+        SnapshotFiles snapshot2 = new SnapshotFiles("1", fileInfos.subList(0, fileInfos.size() - 10), null);
+        return new BlobStoreIndexShardSnapshots(List.of(snapshot1, snapshot2));
+    }
+
+    private static boolean isSame(BlobStoreIndexShardSnapshots shardSnapshot1, BlobStoreIndexShardSnapshots shardSnapshot2) {
```

Comment:
minor, no need to change: You could create a custom `BlobStoreIndexShardSnapshotsAssert` and implement the `isSame()` there.

Reply:
I think custom assertion is nice to have if we re-use it - but so far it the only usage. 

I would leave it as is, if it's fine

Reply:
totally fine.